api/tracer_test.go

```diff
@@ -0,0 +1,118 @@
+package api
+
+import (
+	"fmt"
+	"net/http"
+	"net/http/httptest"
+	"testing"
+
+	"github.com/gobuffalo/uuid"
+	"github.com/netlify/gotrue/conf"
+	"github.com/opentracing/opentracing-go"
+	"github.com/opentracing/opentracing-go/log"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"github.com/stretchr/testify/suite"
+)
+
+type TracerTestSuite struct {
+	suite.Suite
+	API    *API
+	Config *conf.Configuration
+
+	instanceID uuid.UUID
+}
+
+func TestTracer(t *testing.T) {
+	api, config, instanceID, err := setupAPIForTestForInstance()
+	require.NoError(t, err)
+
+	ts := &TracerTestSuite{
+		API:        api,
+		Config:     config,
+		instanceID: instanceID,
+	}
+	defer api.db.Close()
+
+	suite.Run(t, ts)
+}
+
+func (ts *TracerTestSuite) TestTracer_Spans() {
+	spans := []string{}
+
+	tt := &testTracer{spans: &spans, tags: &map[string]string{}}
+	opentracing.SetGlobalTracer(tt)
+
+	w := httptest.NewRecorder()
+	req := httptest.NewRequest(http.MethodPost, "http://localhost/something1", nil)
+	ts.API.handler.ServeHTTP(w, req)
+	req = httptest.NewRequest(http.MethodPost, "http://localhost/something2", nil)
+	ts.API.handler.ServeHTTP(w, req)
+
+	assert.Equal(ts.T(), http.StatusNotFound, w.Code)
+	assert.Equal(ts.T(), spans, []string{"http.handler", "http.handler"})
+}
+
+func (ts *TracerTestSuite) TestTracer_Tags() {
+	tags := map[string]string{}
+
+	tt := &testTracer{tags: &tags, spans: &[]string{}}
+	opentracing.SetGlobalTracer(tt)
+
+	w := httptest.NewRecorder()
+	req := httptest.NewRequest(http.MethodPost, "http://localhost/something", nil)
+	ts.API.handler.ServeHTTP(w, req)
+
+	assert.Equal(ts.T(), http.StatusNotFound, w.Code)
+	assert.Equal(ts.T(), tags["http.method"], "POST")
+	assert.Equal(ts.T(), tags["http.status_code"], "404")
+	assert.Equal(ts.T(), tags["http.url"], "/something")
+	assert.Equal(ts.T(), tags["resource.name"], "POST /something")
+	assert.NotEmpty(ts.T(), tags["http.request_id"])
+}
+
+type testTracer struct {
+	noopTracer opentracing.NoopTracer
+	spans      *[]string
+	tags       *map[string]string
+}
+
+func (tt testTracer) StartSpan(operationName string, opts ...opentracing.StartSpanOption) opentracing.Span {
+	*tt.spans = append(*(tt.spans), operationName)
+	noopSpan := tt.noopTracer.StartSpan(operationName, opts...)
+	return &testSpan{tags: tt.tags, noopSpan: noopSpan}
+}
+func (tt testTracer) Inject(sm opentracing.SpanContext, format interface{}, carrier interface{}) error {
+	return nil
+}
+func (tt testTracer) Extract(format interface{}, carrier interface{}) (opentracing.SpanContext, error) {
+	return tt.noopTracer.Extract(format, carrier)
+}
+
+type testSpan struct {
```

Comment:
@rybit @mraerino Testing the span tags felt a little cumbersome. Would love suggestions if you see a better way.

Reply:
I think that it is good to try and test the span tags you're using. Datadog is very very picky about what tags are there. If they're not you could break the whole APM setup. 

Also look at https://github.com/netlify/netlify-commons/pull/189/files#diff-04def44a4cd6a9c8549a994b1738c32da5163b39623f56abea1b20ae1ea14428 

Specifically I think you can use `github.com/opentracing/opentracing-go/mocktracer` to cut down on some of the boilerplate

Reply:
I was looking for something like `mocktracer`, and not sure why I didn't see it 🤦‍♀️ . Thanks for the pointer.